Delegate calls from VirtualDesktop to the real Desktop when available

mvilliger opened the pull request:
If a ClientRunContext is created (e.g. by using copyCurrent) the values
from the currently active context are copied. The values are the ones
from the moment of the RunContext creation.

E.g. if an async task is scheduled while the Desktop is constructing,
the VirtualDesktop is on the RunContext (as no other Desktop has been
created yet). This means the async task is executed with the
VirtualDesktop as well, even if it is actually executed after the real
Desktop has been created.

This means that the second task uses the VirtualDesktop which has
already been removed from the Session and has been replaced with the
real one. As the VirtualDesktop throws Exceptions for most of the
methods, this may lead to unexpected behavior.

Now the VirtualDesktop gets the real Desktop as soon as it is set to the
session and delegates all calls to it. Older RunContexts still having
the VirtualDesktop will then start to use the real one transparently and
immediately after it is available on the session and without changing
the RunContext or re-scheduling anything. After the real Desktop has
been set to the virtual one, the VirtualDesktop acts like a proxy
delegating all calls to the real one.

As calling methods on the Desktop requires to be in a ModelJob, the
Desktop cannot change during a single execution.

320004

andibur commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/ui/desktop/internal/VirtualDesktop.java:
```diff
@@ -70,1001 +74,1094 @@
+   */
+  public void setRealDesktop(IDesktop realDesktop) {
+    if (realDesktop == null) {
+      throw new IllegalArgumentException("Real desktop must not be virtual");
```
```suggestion
      throw new IllegalArgumentException("Real desktop must not be null");
```

andibur reviewed (approved):
Nice, Thanks

mvilliger committed:
Delegate calls from VirtualDesktop to the real Desktop when available

If a ClientRunContext is created (e.g. by using copyCurrent) the values
from the currently active context are copied. The values are the ones
from the moment of the RunContext creation.

E.g. if an async task is scheduled while the Desktop is constructing,
the VirtualDesktop is on the RunContext (as no other Desktop has been
created yet). This means the async task is executed with the
VirtualDesktop as well, even if it is actually executed after the real
Desktop has been created.

This means that the second task uses the VirtualDesktop which has
already been removed from the Session and has been replaced with the
real one. As the VirtualDesktop throws Exceptions for most of the
methods, this may lead to unexpected behavior.

Now the VirtualDesktop gets the real Desktop as soon as it is set to the
session and delegates all calls to it. Older RunContexts still having
the VirtualDesktop will then start to use the real one transparently and
immediately after it is available on the session and without changing
the RunContext or re-scheduling anything. After the real Desktop has
been set to the virtual one, the VirtualDesktop acts like a proxy
delegating all calls to the real one.

As calling methods on the Desktop requires to be in a ModelJob, the
Desktop cannot change during a single execution.

320004
